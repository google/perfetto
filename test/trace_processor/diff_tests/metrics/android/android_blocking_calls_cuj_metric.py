#!/usr/bin/env python3
# Copyright (C) 2023 The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from os import sys, path
import synth_common

# com.android.systemui
SYSUI_PID = 1000
# com.google.android.apps.nexuslauncher
LAUNCHER_PID = 2000

THIRD_PROCESS_PID = 3000

TOP_LEVEL_SLICES_PID = 4000

# List of blocking calls
blocking_call_names = [
    'monitor contention with something else', 'SuspendThreadByThreadId 123',
    'LoadApkAssetsFd 123', 'binder transaction', 'inflate',
    'Lock contention on thread list lock (owner tid: 1665)',
    "Lock contention on thread suspend count lock (owner tid: 0)",
    "Lock contention on a monitor lock (owner tid: 0)",
    'android.os.Handler: kotlinx.coroutines.CancellableContinuationImpl',
    'relayoutWindow*', 'measure', 'layout', 'configChanged',
    'Contending for pthread mutex', 'ImageDecoder#decodeBitmap',
    'ImageDecoder#decodeDrawable', 'NotificationStackScrollLayout#onMeasure',
    'ExpNotRow#onMeasure(MessagingStyle)', 'ExpNotRow#onMeasure(BigTextStyle)',
    'animation', 'input', 'traversal', 'postAndWait',
    'android.os.Handler: kotlinx.coroutines.internal.DispatchedContinuation',
    'GC: Wait For Completion Alloc', 'Should not be in the metric'
]

top_level_names = [
    'android.view.ViewRootImpl$ViewRootHandler: android.view.View$$Lambda4',
    'android.os.AsyncTask$InternalHandler: #1',
    'android.os.Handler: com.android.systemui.broadcast.ActionReceiver$1$1',
    'com.android.keyguard.KeyguardUpdateMonitor$13: #302',
    'android.os.Handler: com.android.systemui.qs.external.TileServiceManager$1',
    # The following are not expected in the output
    'receiveMessage(inputChannel=62b8bb4 NotificationShade',
    'android.os.Handler: #0',
]


def add_main_thread_atrace(trace, ts, ts_end, buf, pid):
  trace.add_atrace_begin(ts=ts, tid=pid, pid=pid, buf=buf)
  trace.add_atrace_end(ts=ts_end, tid=pid, pid=pid)


def add_async_trace(trace, ts, ts_end, buf, pid):
  trace.add_atrace_async_begin(ts=ts, tid=pid, pid=pid, buf=buf)
  trace.add_atrace_async_end(ts=ts_end, tid=pid, pid=pid, buf=buf)


def add_binder_transaction(trace, tx_pid, rx_pid, start_ts, end_ts):
  trace.add_binder_transaction(
      transaction_id=tx_pid,
      ts_start=start_ts,
      ts_end=end_ts,
      tid=tx_pid,
      pid=tx_pid,
      reply_id=rx_pid,
      reply_ts_start=start_ts,
      reply_ts_end=end_ts,
      reply_tid=rx_pid,
      reply_pid=rx_pid)


# Adds a set of predefined blocking calls in places near the cuj boundaries to
# verify that only the portion inside the cuj is counted in the metric.
def add_cuj_with_blocking_calls(trace, cuj_name, pid):
  cuj_begin = 2_000_000
  cuj_dur = 15_000_000
  cuj_end = cuj_begin + cuj_dur
  blocking_call_name = "binder transaction"

  add_async_trace(trace, ts=cuj_begin, ts_end=cuj_end, buf=cuj_name, pid=pid)

  # all outside, before cuj, shouldn't be counted
  add_main_thread_atrace(
      trace,
      ts=cuj_begin - 2_000_000,
      ts_end=cuj_begin - 1_000_000,
      buf=blocking_call_name,
      pid=pid)

  # mid inside, mid outside. Should account for half the time.
  add_main_thread_atrace(
      trace,
      ts=cuj_begin - 1_000_000,
      ts_end=cuj_begin + 1_000_000,
      buf=blocking_call_name,
      pid=pid)

  # completely inside
  add_main_thread_atrace(
      trace,
      ts=cuj_begin + 2_000_000,
      ts_end=cuj_begin + 3_000_000,
      buf=blocking_call_name,
      pid=pid)

  add_main_thread_atrace(
      trace,
      ts=cuj_begin + 4_000_000,
      ts_end=cuj_begin + 7_000_000,
      buf=blocking_call_name,
      pid=pid)

  # mid inside, mid outside
  add_main_thread_atrace(
      trace,
      ts=cuj_end - 1_000_000,
      ts_end=cuj_end + 1_000_000,
      buf=blocking_call_name,
      pid=pid)

  # all outside, after cuj, shouldn't be counted/
  add_main_thread_atrace(
      trace,
      ts=cuj_end + 2_000_000,
      ts_end=cuj_end + 3_000_000,
      buf=blocking_call_name,
      pid=pid)


def add_cuj_with_top_level_slices(trace, cuj_name, pid):
  blocking_call_dur = 10_000_000
  blocking_call_ts = 2_000_000

  cuj_dur = len(top_level_names) * blocking_call_dur
  add_async_trace(
      trace,
      ts=blocking_call_ts,
      ts_end=blocking_call_ts + cuj_dur,
      buf=cuj_name,
      pid=pid)

  for top_level_slice in top_level_names:
    add_main_thread_atrace(
        trace,
        ts=blocking_call_ts,
        ts_end=blocking_call_ts + blocking_call_dur,
        buf=top_level_slice,
        pid=pid)
    blocking_call_ts += blocking_call_dur

  # Some top level unrelated to handler
  add_main_thread_atrace(
      trace,
      ts=blocking_call_ts,
      ts_end=blocking_call_ts + blocking_call_dur,
      buf="some top level slice that should not be in the output",
      pid=pid)
  # Nested inside the previous, should not be in the output as not top level.
  add_main_thread_atrace(
      trace,
      ts=blocking_call_ts + 1,
      ts_end=blocking_call_ts + blocking_call_dur - 1,
      buf="should.not.be.in.the.output.Handler: not.in.the.output$1",
      pid=pid)


# Creates a cuj that contains one of each blocking call.
def add_all_blocking_calls_in_cuj(trace, pid):
  blocking_call_dur = 10_000_000
  blocking_call_ts = 2_000_000

  cuj_dur = len(blocking_call_names) * blocking_call_dur
  add_async_trace(
      trace,
      ts=blocking_call_ts,
      ts_end=blocking_call_ts + cuj_dur,
      buf="L<CUJ_WITH_MANY_BLOCKING_CALLS>",
      pid=pid)

  for blocking_call in blocking_call_names:
    add_main_thread_atrace(
        trace,
        ts=blocking_call_ts,
        ts_end=blocking_call_ts + blocking_call_dur,
        buf=blocking_call,
        pid=pid)
    blocking_call_ts += blocking_call_dur


# Creates 2 overlapping cuj, and a blocking call that lasts for both of them.
def add_overlapping_cujs_with_blocking_calls(trace, start_ts, pid):
  add_async_trace(
      trace,
      ts=start_ts,
      ts_end=start_ts + 10_000_000,
      buf="L<OVERLAPPING_CUJ_1>",
      pid=pid)
  add_async_trace(
      trace,
      ts=start_ts + 2_000_000,
      ts_end=start_ts + 12_000_000,
      buf="L<OVERLAPPING_CUJ_2>",
      pid=pid)

  add_main_thread_atrace(
      trace,
      ts=start_ts,
      ts_end=start_ts + 12_000_000,
      buf=blocking_call_names[0],
      pid=pid)


def add_cuj_with_named_binder_transaction(pid, rx_pid):
  cuj_begin = 40_000_000
  cuj_end = 50_000_000

  add_async_trace(
      trace,
      ts=cuj_begin,
      ts_end=cuj_end,
      buf="L<WITH_NAMED_BINDER_TRANSACTION>",
      pid=pid)

  add_binder_transaction(
      trace, tx_pid=pid, rx_pid=rx_pid, start_ts=cuj_begin, end_ts=cuj_end)

  # Slice inside the binder reply, to give a name to the binder call.
  # The named binder slice introduced should be the length of the entire
  # transaction even if the "name" slice only covers some of the binder reply.
  add_main_thread_atrace(
      trace,
      ts=cuj_begin + 1_000_000,
      ts_end=cuj_end - 1_000_000,
      buf="AIDL::java::IWindowManager::hasNavigationBar::server",
      pid=rx_pid)


def add_process(trace, package_name, uid, pid):
  trace.add_package_list(ts=0, name=package_name, uid=uid, version_code=1)
  trace.add_process(pid=pid, ppid=pid, cmdline=package_name, uid=uid)
  trace.add_thread(tid=pid, tgid=pid, cmdline="MainThread", name="MainThread")


def setup_trace():
  trace = synth_common.create_trace()
  trace.add_packet()
  add_process(
      trace, package_name="com.android.systemui", uid=10001, pid=SYSUI_PID)
  add_process(
      trace,
      package_name="com.google.android.apps.nexuslauncher",
      uid=10002,
      pid=LAUNCHER_PID)
  add_process(
      trace,
      package_name="com.google.android.third.process",
      uid=10003,
      pid=THIRD_PROCESS_PID)
  add_process(
      trace,
      package_name="com.google.android.top.level.slices",
      uid=10004,
      pid=TOP_LEVEL_SLICES_PID)
  trace.add_ftrace_packet(cpu=0)
  add_async_trace(trace, ts=0, ts_end=5, buf="J<IGNORED>", pid=SYSUI_PID)
  return trace


trace = setup_trace()

add_cuj_with_blocking_calls(trace, "L<TEST_SYSUI_LATENCY_EVENT>", pid=SYSUI_PID)
add_cuj_with_blocking_calls(
    trace, "L<TEST_LAUNCHER_LATENCY_EVENT>", pid=LAUNCHER_PID)
add_cuj_with_top_level_slices(
    trace, "L<CUJ_WITH_TOP_LEVEL_SLICES>", pid=TOP_LEVEL_SLICES_PID)

add_all_blocking_calls_in_cuj(trace, pid=THIRD_PROCESS_PID)

add_overlapping_cujs_with_blocking_calls(
    trace, pid=SYSUI_PID, start_ts=20_000_000)

add_cuj_with_named_binder_transaction(pid=SYSUI_PID, rx_pid=LAUNCHER_PID)

# Note that J<*> events are not tested here.
# See test_android_blocking_calls_on_jank_cujs.
sys.stdout.buffer.write(trace.trace.SerializeToString())
