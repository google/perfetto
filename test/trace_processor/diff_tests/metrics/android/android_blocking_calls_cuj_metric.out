android_blocking_calls_cuj_metric {
  cuj {
    id: 1
    name: "TEST_SYSUI_LATENCY_EVENT"
    process {
      name: "com.android.systemui"
      uid: 10001
      pid: 1000
      android_user_id: 0
      is_kernel_task: false
    }
    ts: 2000000
    dur: 15000000
    blocking_calls {
      name: "binder transaction"
      cnt: 4
      total_dur_ms: 6
      max_dur_ms: 3
      min_dur_ms: 1
      total_dur_ns: 6000000
      max_dur_ns: 3000000
      min_dur_ns: 1000000
    }
  }
  cuj {
    id: 2
    name: "TEST_LAUNCHER_LATENCY_EVENT"
    process {
      name: "com.google.android.apps.nexuslauncher"
      uid: 10002
      pid: 2000
      android_user_id: 0
      is_kernel_task: false
    }
    ts: 2000000
    dur: 15000000
    blocking_calls {
      name: "binder transaction"
      cnt: 4
      total_dur_ms: 6
      max_dur_ms: 3
      min_dur_ms: 1
      total_dur_ns: 6000000
      max_dur_ns: 3000000
      min_dur_ns: 1000000
    }
  }
  cuj {
    id: 3
    name: "CUJ_WITH_MANY_BLOCKING_CALLS"
    process {
      name: "com.google.android.third.process"
      uid: 10003
      pid: 3000
      android_user_id: 0
      is_kernel_task: false
    }
    ts: 2000000
    dur: 260000000
    blocking_calls {
      name: "Contending for pthread mutex"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "CoroutineContinuation"
      cnt: 2
      total_dur_ms: 20
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 20000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "ExpNotRow#onMeasure(BigTextStyle)"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "ExpNotRow#onMeasure(MessagingStyle)"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "Garbage Collector"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "ImageDecoder#decodeBitmap"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "ImageDecoder#decodeDrawable"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "LoadApkAssetsFd <...>"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "Lock contention on a monitor lock <...>"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "Lock contention on thread list lock <...>"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "Lock contention on thread suspend count lock <...>"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "NotificationStackScrollLayout#onMeasure"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "SuspendThreadByThreadId <...>"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "animation"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "binder transaction"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "configChanged"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "inflate"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "input"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "layout"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "measure"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "monitor contention with <...>"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "postAndWait"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "relayoutWindow <...>"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "traversal"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
  }
  cuj {
    id: 4
    name: "CUJ_WITH_TOP_LEVEL_SLICES"
    process {
      name: "com.google.android.top.level.slices"
      uid: 10004
      pid: 4000
      android_user_id: 0
      is_kernel_task: false
    }
    ts: 2000000
    dur: 70000000
    blocking_calls {
      name: "Handler: android.os.AsyncTask"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "Handler: android.view.View"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "Handler: com.android.keyguard.KeyguardUpdateMonitor"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "Handler: com.android.systemui.broadcast.ActionReceiver"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "Handler: com.android.systemui.qs.external.TileServiceManager"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
  }
  cuj {
    id: 5
    name: "OVERLAPPING_CUJ_1"
    process {
      name: "com.android.systemui"
      uid: 10001
      pid: 1000
      android_user_id: 0
      is_kernel_task: false
    }
    ts: 20000000
    dur: 10000000
    blocking_calls {
      name: "monitor contention with <...>"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
  }
  cuj {
    id: 6
    name: "OVERLAPPING_CUJ_2"
    process {
      name: "com.android.systemui"
      uid: 10001
      pid: 1000
      android_user_id: 0
      is_kernel_task: false
    }
    ts: 22000000
    dur: 10000000
    blocking_calls {
      name: "monitor contention with <...>"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
  }
  cuj {
    id: 7
    name: "WITH_NAMED_BINDER_TRANSACTION"
    process {
      name: "com.android.systemui"
      uid: 10001
      pid: 1000
      android_user_id: 0
      is_kernel_task: false
    }
    ts: 40000000
    dur: 10000000
    blocking_calls {
      name: "AIDL::java::IWindowManager::hasNavigationBar::server"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
    blocking_calls {
      name: "binder transaction"
      cnt: 1
      total_dur_ms: 10
      max_dur_ms: 10
      min_dur_ms: 10
      total_dur_ns: 10000000
      max_dur_ns: 10000000
      min_dur_ns: 10000000
    }
  }
}
